### PR TITLE
[WIP] BoxedResidue: fix constant-time TODO in `montgomery_mul`

### PR DESCRIPTION
**Pull request by tarcieri**
Replaces branching with an arithmetic-based approach. This unfortunately seems to double the time multiplication takes (and with it, modpow).

### Benchmarks

```
Montgomery arithmetic/multiplication, BoxedUint*BoxedUint
                        time:   [10.027 µs 10.048 µs 10.068 µs]
                        change: [+101.70% +102.44% +103.19%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

```
Montgomery arithmetic/modpow, BoxedUint^BoxedUint
                        time:   [48.200 ms 48.273 ms 48.352 ms]
                        change: [+97.268% +97.613% +97.972%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

### CODE REVIEW

File: src/modular/boxed_residue/mul.rs [hunk 1]
```diff
@@ -266,3 +265,9 @@ fn mul_add_www(x: Limb, y: Limb, c: Limb) -> (Limb, Limb) {
     let z = x.0 as WideWord * y.0 as WideWord + c.0 as WideWord;
     (Limb((z >> Word::BITS) as Word), Limb(z as Word))
 }
+
+/// Compare limbs in constant time, returning `Limb::ONE` if the left size is less than the right.
+#[inline(always)]
+fn limb_ct_lt(a1: Limb, b1: Limb, a2: Limb, b2: Limb) -> Limb {
+    (a1.sbb(b1, Limb::ZERO).1 | a2.sbb(b2, Limb::ZERO).1) & Limb::ONE
+}
```
Comment (tarcieri) [inline src/modular/boxed_residue/mul.rs:273]: This was faster than using the `ct_lt` approach previously used on L214, but still slower than using branch instructions

**Comment by fjarri** [comment]
This could benefit from #418

**Comment by fjarri** [comment]
Doesn't seem like it benefits. 

But judging by Godbolt, using `overflowing_add()` there doesn't lead to branches. Directly casting the booleans to `Word` produces the same code as the usage of `overflowing_add()`.

**Comment by tarcieri** [comment]
Closing this for now

**Pull request closed without merging**